### ISSUE TITLE
ShMemIterator should be RandomAccessIterator

### ISSUE THREAD

**Issue by daniel-j-h** [Project-OSRM/osrm-backend#3458]
The ShM wrapper has either a vector or a shared mem block as underlying storage. It makes no sense for its `begin()` and `end()` functions to return an `InputIterator` which you can't use for random access, and can't e.g. subtract from another.

https://github.com/Project-OSRM/osrm-backend/blob/9832825abd4f55403ae3fcaa006ad3f9b503687a/include/util/shared_memory_vector_wrapper.hpp#L21-L47

@danpat you played with ripping the wrapper out once and for all, what was the progress there? As it is now, the iterator interface is completely useless and even dangerous: e.g. `std::distance` dispatches on the iterator tag and would loop over all elements for counting: O(n) instead of O(1).

**Comment by danpat** [comment]
I played with `boost::iterator_range` which seems to work as a pretty-much drop-in replacement for *most* cases where we've got our `shared_memory_vector_wrapper` right now.  However, I pulled that https://github.com/Project-OSRM/osrm-backend/pull/3165 in order to not clutter the change with too many things.

We should re-visit this - using `boost::iterator_range` was trivial in most cases.  There were a few sticking points, but the bulk was mostly a search/replace exercise.
